Feature/3049 coauthorable Proposals

tramuntanal opened the pull request:
#### :tophat: What? Why?
As in the near future Proposals mey be possibly produced from CollaborativeDrafts with multiple co-authors, this PR prepares Proposals to support coauthorship.

This PR won't change any workflow, only will refactor Proposals to support coauthorship.

#### :pushpin: Related Issues
- Related to #3049 

#### :clipboard: Subtasks
- [ ] Add `CHANGELOG` entry
- [x] Add documentation regarding the feature 
- [x] Modify Proposals seeds to use coauthorsihp instead of authorship.
- [x] Modify tests
- [x] Modify test Factories
- [x] Migrate current Proposal authorship to Coauthorship.

tramuntanal commented:
Hi @decidim/lot-core,
We're having the same decidim-generators problem here (other failing tests are pending). I can say that this commit 5643a782b80b82dbd1bf5196a08d5ecc13b4d69e contains this:
```
+# frozen_string_literal: true
+
+class MoveAuthorshipsToCoauthorships < ActiveRecord::Migration[5.1]
+  class Proposal < ApplicationRecord
+    self.table_name = :decidim_proposals_proposals
+  end
+  class Coauthorship < ApplicationRecord
+    self.table_name = :decidim_coauthorships
+  end
+
+  def change
+    proposals = Proposal.all
+
+    proposals.each do |proposal|
+      author_id = proposal.attributes["decidim_author_id"]
+      user_group_id = proposal.attributes["decidim_user_group_id"]
+      Decidim::Coauthorship.create!(
+        coauthorable_id: proposal.id,
+        coauthorable_type: "Decidim::Proposals::Proposal",
+        decidim_author_id: author_id,
+        decidim_user_group_id: user_group_id
+      )
+    end
+  end
+end
```

As you can see I forgot to remove the `Decidim::` prefix for Coauthorship. I suspect that this is causing the decidim-generators problem: invoking models from migrations. Do you know how this is affecting the specs in generators module?

mrcasals commented:
@tramuntanal I have no idea sincerely, maybe @deivid-rodriguez can help us here? 😄

deivid-rodriguez commented:
In the generator spec that seeds, I guess seeding is failing. However, I remember improving error messages on these specs so the error should have been more clear... I guess that stuff is not playing nice with the junit formatter.

Can you try running the specs without the fix and without the formatter, and see if it fails more "clearly"?

tramuntanal commented:
After the refactor decidim-generators still fails. So this was not the cause although in https://github.com/decidim/decidim/pull/3559 changing the use of a fake class instead of the model solved the problem.

@deivid-rodriguez I removed the JUnit formatter and replaced with doc formatter and it is still failing, it seems that it has nothing to do with the format. Also, now it is not rendering anything, it seems it is outputting to some stdout that circle is not redirecting :(

deivid-rodriguez commented:
Not sure why circleCI swallows the output but locally the bug seems pretty clear:

```
       Creating seeds for decidim-core...
       Creating seeds for decidim-system...
       Creating seeds for the participatory_processes space...
       -- Creating pages component seeds for the participatory space with ID: 1...
       -- Creating meetings component seeds for the participatory space with ID: 1...
       rails aborted!
       ActiveModel::UnknownAttributeError: unknown attribute 'author' for Decidim::Proposals::Proposal.
       /home/deivid/Code/decidim/decidim-proposals/lib/decidim/proposals/component.rb:140:in `block (3 levels) in <top (required)>'
       /home/deivid/Code/decidim/decidim-proposals/lib/decidim/proposals/component.rb:127:in `times'
       /home/deivid/Code/decidim/decidim-proposals/lib/decidim/proposals/component.rb:127:in `block (2 levels) in <top (required)>'
       /home/deivid/Code/decidim/decidim-core/lib/decidim/component_manifest.rb:103:in `seed!'
       /home/deivid/Code/decidim/decidim-participatory_processes/lib/decidim/participatory_processes/participatory_space.rb:155:in `block (4 levels) in <top (required)>'
       /home/deivid/Code/decidim/decidim-participatory_processes/lib/decidim/participatory_processes/participatory_space.rb:154:in `block (3 levels) in <top (required)>'
       /home/deivid/Code/decidim/decidim-participatory_processes/lib/decidim/participatory_processes/participatory_space.rb:52:in `times'
       /home/deivid/Code/decidim/decidim-participatory_processes/lib/decidim/participatory_processes/participatory_space.rb:52:in `block (2 levels) in <top (required)>'
       /home/deivid/Code/decidim/decidim-core/lib/decidim/participatory_space_manifest.rb:83:in `seed!'
       /home/deivid/Code/decidim/decidim-core/lib/decidim/core.rb:75:in `seed!'
       /home/deivid/Code/decidim/decidim-generators/spec/generator_test_app/db/seeds.rb:9:in `<top (required)>'
       bin/rails:4:in `<main>'
       Tasks: TOP => db:seed
       (See full trace by running task with --trace)
       -- Creating proposals component seeds for the participatory space with ID: 1...
     Shared Example Group: "a sane generator" called from ./spec/generators_spec.rb:90
     # ./spec/generators_spec.rb:27:in `block (3 levels) in <module:Decidim>'

Finished in 5 minutes 8 seconds (files took 0.15229 seconds to load)
6 examples, 1 failure

Failed examples:

rspec ./spec/generators_spec.rb[1:1:5:1:1] # Decidim::Generators with an application with a development application behaves like a sane generator successfully generates application
```

deivid-rodriguez commented:
Ok, I see what's going on in circle now. Fix coming.

deivid-rodriguez commented:
@tramuntanal If you rebase or merge master into this branch, you should get a clear failure in CI now.

tramuntanal commented:
Hi @decidim/lot-core,
Still with the decidim-generators problem. After switching format to "doc" I can see [here](https://circleci.com/gh/decidim/decidim/112721?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) that the spec that is not executing (or finishinig, or crashing) is "with an application with a development application behaves like a sane generator" and it lacks the "successfully generates application" part, so it seems that something strange happens here and this tests does not finishes, the next spec does not executes, rspec does not tell what happened but exists with code 1.

I'm runnig this specific spec locally and it happens exactly the same, no "successfully generates application" part, but also no output of any kind. After adding some traces I get:
```
  with an application
    with a development application
      behaves like a sane generator
OUT: /home/oliver/.rbenv/versions/2.5.1/lib/ruby/2.5.0/rubygems/dependency.rb:283:in `matching_specs': uninitialized constant Gem::BundlerVersionFinder (NameError)
Did you mean?  Gem::BundlerVersionFinder
	from /home/oliver/.rbenv/versions/2.5.1/lib/ruby/2.5.0/rubygems/dependency.rb:302:in `to_specs'
```
Let's add this kind of traces when GemManager fails to execute a command.

josepjaume commented:
Nicely done! Let's see if we can figure this out!

tramuntanal commented:
Got it!
After adding logs GemManager is not silent anymore, and now we can see why processes are aborting, the culprit is a problem in the seeds: https://circleci.com/gh/decidim/decidim/113115?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Easy to solve, but the lack of error message made it very hard to find here and in other PRs.

@decidim/lot-core After adding this traces changes in common_rake.rb can be reviewed. Do you agree in keeping with the "doc" format? Do you need JUnit format? We can keep both or only the first, I would remove JUnit if you confirm that it is not used for anything.

mrcasals commented:
@tramuntanal the JUnit format is used for coverage, so we need it.

tramuntanal commented:
Ok @mrcasals we have this options now: `--format doc --format RspecJunitFormatter -o ~/rspec/rspec.xml`.

deivid-rodriguez commented:
@tramuntanal have you seen my messages and related PR? I'm not sure what the problem is now... The gem manager does not need any logging since the specs output the whole thing in case of errors. Also I copied the error in seeds yesterday into this thread. I'm confused :thinking:

tramuntanal commented:
Hi @deivid-rodriguez, I read the comment but i thought you were referring to the --fail-fast option, not that you were chaning the implementation of the GemManager. :thinking: 

Now I see your modification, I don't think that fail-fast should be used because we will get only one error if there are many specs failing. Why did you add this flag?

deivid-rodriguez commented:
Fail fast has been enabled in CI for ages, in order to be a bit faster and let failing builds use less resources. This personally works well for me since I normally run tests locally before opening PR's. But if you use CI as a substitute for running tests locally, then I agree fail fast is not good, because on heavy WIP branches you'll want full test feedback. 

To summarize, I don't mind switching this option but bear in mind that it will make your CI bottlenecks worse.

mrcasals commented:
I'd keep the fail fast option, sincerely. Tests are slow enough already, if we make the bottleneck harder then it'll become a hell.

tramuntanal commented:
@deivid-rodriguez then you run the whole test suite locally before every push?
It takes more than 1 hour, if I have many PR's open in my laptop I can not wait more than 1hour to switch between branches. I already have 3 decidim installations in my computer only for development but still there are problems, for example when I need to clean the dummy_app with `rake test_app` it generates the database file with the defaults and then breaks the tests running in the installation that uses the defaults. Not to say that the computer becomes slow if you run the test suite, imagine running 2 test suites in parallel.

Maybe I'm doing it wrong here but I feel it is unpractical to run whole tests locally. This is one thing to improve.

On the other side, if we have bottlenecks with circleci workers, ok let's keep the --fail-fast option. It will at least make us notice there is a problem in one module that must be checked locally.

mrcasals commented:
@tramuntanal my usual workflow is:

1. Write the code
2. Run the tests I think that can be affected locally, and fix the issues
3. Push to GitHub to get the CI builds
4. Check other modules that might fail
5. Fix those modules locally. If needed, run all tests locally for those modules to be sure
6. Repeat step 3 until CI is green

I try to minimize the number of CI builds, because they block everyone else.

deivid-rodriguez commented:
Exactly, that's what I do too. I run the tests that I think could be affected (or the ones I'm developing) locally, not everything of course.

Regarding test apps, I'm not sure about the problem you're talking about. I think I remember from other PR's that you manually edit the database configuration file on generated test applications, maybe that's related. I work with several repos at the same time and I don't have any problems with test apps... :thinking:

tramuntanal commented:
@deivid-rodriguez when you create a test_app or a development_app (the problem is with both) the name of the database is a default one, I think `decidim_development_app_development`.

If I locally clone many Decidim, and for each one I generate a test_app, all of the cloned Decidims will share the same test (or devel) database. To avoid problems with migrations I rename the database for each of the clones.

deivid-rodriguez commented:
But I think the first `decidim_` is taken from the base folder, so it would only conflict if you name all of your decidim clones with the same name. Isn't that very confusing? How do you distinguish which clone are you working on? Also, why do you work on several clones of the same repo? Is it to be able to run tests in parallel locally or something? :thinking:

tramuntanal commented:
@deivid-rodriguez I gess it is not as you explain because I had the same problem when cloning into repositories with different names. Also it wouldn't work because decidim-generators fails if Decidim is not in a folder called `decidim`.

I tried to execute `$ bundle exec rake development_app --app_name=decidim3` but it does not accept the "app_name" parameter.

tramuntanal commented:
@decidim/lot-core This PR is ready to be merged.

deivid-rodriguez commented:
I just tried it and it does exectly what I expect. I cloned decidim into a folder called `decidim_clone`, and the inferred db name was `decidim_clone_development_app_development`. It didn't work on the very first try because of the postgreSQL configuration I personally use, but that's unrelated. You can see the full console log [here](https://gist.github.com/deivid-rodriguez/d0a0c5ca2ae6a33c7396d904d67e06a0) in case it helps you.

tramuntanal commented:
@deivid-rodriguez you are right! it is strange because I remember having to manually change the database name, maybe it was after restructuring directories due to decidim-generators problem. Just this problem impedes having clonning Decidim into a directory that is not named `decidim` because otherwise all generators test fail like:
```
  1) Decidim::Generators with an application without flags behaves like a sane generator successfully generates application
     Failure/Error: expect(result[1]).to be_success, result[0]
     
       rbenv: decidim: command not found
     
       The `decidim' command exists in these Ruby versions:
         2.5.0
     Shared Example Group: "a sane generator" called from ./spec/generators_spec.rb:67
     # ./spec/generators_spec.rb:28:in `block (3 levels) in <module:Decidim>'
....
```

deivid-rodriguez commented:
Well, the fix so that this works is relatively recent, so it was probably broken when you last tried it. 

Regarding your other problem, it seems like a bug indeed! It should be a simple fix though! :+1:

tramuntanal commented:
By the way, @decidim/lot-core , it is not urgent for 0.12 but when you have the time this PR is required to have #3049 for 0.13

mrcasals commented on decidim-core/app/cells/decidim/collapsible_list_cell.rb:
```diff
@@ -19,7 +19,7 @@ module Decidim
   #      size: :small
   #    )
-  class CollapsibleListCell < Decidim::ViewModel
+  class CollapsibleListCell < AuthorCell
```
This cell is meant to list any kind of things. Making it inherit from `Decidim::AuthorCell` only lets us show authors, and you're coupling it with the `render :withdraw` part in the view side. 

Can we find another way to achieve the same result?

mrcasals commented on decidim-core/lib/decidim/coauthorable.rb:
```diff
@@ -0,0 +1,64 @@
+      #
+      # @param user: The new coauthor.
+      # @extra_attrs: Extra
+      def add_coauthor(user, extra_attrs = {})
```
This logic should probably belong to a command

tramuntanal replied:
This method is used inside the corresponding create and update commands and in seeds I think this is its right place.

tramuntanal replied:
Having this method here is:
- a way so that Create and Update commands can reuse code, both have to add coauthors.
- syntactic sugar so that commands and seeds can create objects in one line. This method name has more meaning than `Decidim::Coauthorship.create!(user_group: user_group)`
- add_coauthor is a task already tackled by these create and update commands, why add a sub-command empty of logic if is not needed?

By the way, I see that there are already other interfaces with this kind of features like [Resourceable#link_resources](https://github.com/decidim/decidim/blob/58a1ba7c3a16dfc3643ca8e459eca3497ffb83e8/decidim-core/lib/decidim/resourceable.rb#L61) so what I'm doing here is nothing new in the application.

mrcasals replied:
Ok! 👍

mrcasals commented on CHANGELOG.md:
```diff
@@ -9,6 +9,43 @@
 **Changed**:
 
+- **decidim-core**: Introduce coauthorable concern and coauthorship model. [\#3310](https://github.com/decidim/decidim/pull/3310)
+- **decidim-core**: Force user_group.name uniqueness in user_group test factory. (https://github.com/decidim/decidim/pull/3290)
```
This changelog seems wrong

tramuntanal replied:
Sorry @mrcasals you mean the first at line 12 or the second at line 13? I guess it is the second that I have added by error correct?

mrcasals commented on decidim-core/lib/decidim/authorable.rb:
```diff
@@ -1,7 +1,9 @@
-  # This concern contains the logic related to authorship
+  # This concern contains the logic related to the authorship.
+  # This authorship may belong to a single user or be shared among coauthors, in
+  # this latest case the Coauthorable concern should be also applied.
```
So `Coauthorable` needs the `Authorable` concern? Why? also, if it's really needed, we should probably note this in the `Coauthorable` docs.

tramuntanal replied:
No, sorry, this text is deprecated. I'm updating it.

mrcasals commented on decidim-core/lib/decidim/coauthorable.rb:
```diff
@@ -0,0 +1,64 @@
+        authors.first
+      end
+
+      # Syntactic sugar to access first coauthor
```
Wrong docs?

mrcasals commented on decidim-core/lib/decidim/coauthorable.rb:
```diff
@@ -0,0 +1,64 @@
+
+      # Syntactic sugar to access first coauthor
+      def creator_user_group
+        user_groups.first
```
This logic seems wrong, the first user_group coauthoring a proposal might not be the creator of the proposal (could be the 4th coauthor in order, but the first that is a user_group)

tramuntanal replied:
The problem if the name of the method not what it returns as what I wanted was the first user group.
However as it is for testing purposes so I'm refactoring by removing this method and using a simple `#creator` that instead returns the `Coauthorship`.

mrcasals commented on decidim-core/lib/decidim/coauthorable.rb:
```diff
@@ -0,0 +1,64 @@
+  #
+  # Coauthorable shares nearly the same object interface as Authorable but with some differences.
+  # - `authored_by?(user)` is exactly the same
+  # - `normalized_authors` is pluralized
```
This method does not seem to exist?

tramuntanal replied:
Again, this text needs to be updated.

mrcasals commented on decidim-core/lib/decidim/coauthorable.rb:
```diff
@@ -0,0 +1,64 @@
+      has_many :authors, through: :coauthorships, class_name: "Decidim::User"
+      has_many :user_groups, as: :coauthorable, class_name: "Decidim::UserGroup", through: :coauthorships
+
+      scope :from_author, ->(author) { joins(:coauthorships).where("decidim_coauthorships.decidim_author_id": author.id) }
```
We should probably check that the `decidim_user_group_id` column is null

tramuntanal replied:
It is correct this way. But I'll add some documentation to make it clear that it retrieves models from all identities of the user.

tramuntanal replied:
Now I see in ProposalsController that it also has the behaviour you defined so I'll add a new method to disambiguate.

mrcasals replied:
Nice, thanks!

mrcasals commented on decidim-core/lib/decidim/coauthorable.rb:
```diff
@@ -0,0 +1,64 @@
+      #
+      # user - the user to check for authorship
+      def authored_by?(user)
+        coauthorships.where(author: user).or(coauthorships.where(user_group: user.user_groups)).exists?
```
Same about the `decidim_user_group_id` column in the first part of the query

tramuntanal replied:
If I leave the first part as it is now, I even guess if the second part of the query is even required, let me check!

mrcasals commented on decidim-dev/lib/decidim/dev/common_rake.rb:
```diff
@@ -5,7 +5,7 @@
 
 RSpec::Core::RakeTask.new(:spec) do |t|
-  t.rspec_opts = "--format progress --format RspecJunitFormatter -o ~/rspec/rspec.xml" if ENV["CI"]
+  t.rspec_opts = "--format doc --format RspecJunitFormatter -o ~/rspec/rspec.xml" if ENV["CI"]
```
Move this back to the old value?

tramuntanal replied:
It was useful when decidim-generators were lacking feedback on failure and we were not able to know which test failed.
I can re-set `progress` if you want, I see that now circleci is generating very long logs. We can always re-set doc format if we need.
Ok, I'm applying the change, no need to respond.

mrcasals commented on decidim-proposals/app/presenters/decidim/proposals/proposal_presenter.rb:
```diff
@@ -12,10 +12,13 @@ class ProposalPresenter < SimpleDelegator
-                      Decidim::UserPresenter.new(super)
+                      coauthorship = coauthorships.first
+                      if coauthorship.user_group
+                        Decidim::UserGroupPresenter.new(user_group)
```
`coauthorship.user_group`?

tramuntanal replied:
Sorry, I don't get your comment..

tramuntanal replied:
Coauthorable gives access to couauthorships but also direct access to authors and to user_groups.

tramuntanal replied:
You are totally right @mrcasals . Sorry, I was totally misunderstanding what you meant. It should be `coauthorship.user_group`. Fixing it right now!

mrcasals commented on decidim-proposals/app/helpers/decidim/proposals/application_helper.rb:
```diff
@@ -83,6 +83,30 @@ def follow_button_for(model)
+        coauthorable.identities.map { |identity| present(identity) }
+      end
+
+      def proposal_coauthors_cell(proposal, _size = :small)
```
Move this logic into a cell

mrcasals commented on decidim-proposals/db/migrate/20180530131323_add_counter_cache_coauthorships_to_proposals.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class AddCounterCacheCoauthorshipsToProposals < ActiveRecord::Migration[5.1]
```
Merge with `MoveAuthorshipsToCoauthorships` so the cou ter is correctly set

josepjaume commented:
As a basis for a future refactor, I really think this could have been solved by having all the logic inside `Decidim::Authorable`, with a configurable flag. Something like:

```ruby
include Decidim::Authorable
authorable coauthors: true
```

We could then turn `author_id` into something like `author_ids`, mimicking the current `author` interface when `coauthors` is false.

If you agree with that, we can create an issue to refactor this on the future.

tramuntanal commented:
I was just having this same feeling while applying the review changes!
Ok, let's do like you say.

tramuntanal commented:
Ready to merge @decidim/lot-core

tramuntanal commented:
@decidim/lot-core is there anything pending to rework? I think everything is ok isn't it?

agustibr commented:
@decidim/lot-core :information_desk_person: The Collaborative draft MVP PR #3109 depends on this PR, is this going to be merged soon?  :blush:

agustibr commented on decidim-proposals/app/cells/decidim/proposals/coauthorships_cell.rb:
```diff
@@ -0,0 +1,44 @@
+
+      private
+
+      def proposal
```
I don't see the point of this method, as `model` works fine and is more abstract, since in the near future `coauthorships_cell` will be used also in `collaborative_drafts` also.
So I would `s/proposal/model`

mrcasals commented:
@agustibr @tramuntanal left some comments there that I think need to be tackled before merging! 😄

agustibr commented:
OK @mrcasals :satisfied:

tramuntanal commented:
@decidim/lot-core I guess there were only 2 comments that I've answered just now.

mrcasals commented:
Wohoo! 👏

agustibr commented:
Yay! :smile_cat:

tramuntanal commented:
iupiiii!!! :bangbang: :clap:
